serving/purgatorio/select-and-train-machine-learning-models/tools-and-libraries/tools-and-libraries.md

```diff
@@ -1,5 +1,321 @@
+
+- You'll always be capable of learning more libraries when you reach the capabilities limits of the one that you already know or you have different needs.  
+
+Now you know what is the best approach to learn tools and libraries and frameworks, and how to do it [using wisely your time](https://waitbutwhy.com/2014/05/life-weeks.html)!
+
+The following links expand these points and you should read them if you are a beginner, some answers are illuminating:
+
+- [What are some tips for a beginning ML/data scientist who feels overwhelmed?](https://www.quora.com/What-are-some-tips-for-a-beginning-ML-data-scientist-who-feels-overwhelmed)
+- [SO MUCH to Learn... Overwhelmed. How to Learn so Much?](https://www.reddit.com/r/learnprogramming/comments/dnujl0/so_much_to_learn_overwhelmed_how_to_learn_so_much/)
+- [Beginner to "data science" but unsure of what I should be doing](https://www.reddit.com/r/datascience/comments/7yg61i/beginner_to_data_science_but_unsure_of_what_i/)
+
+Having that said, let's jump in!
+
+# The Data Science Process Toolkit
+## Data Collection
+
+Data collection is one of the most crucial parts of a Data Science project (you don't say?), and often a tedious one if you have to label your data on your own!
+
+Luckily, good researchers and institutions often release their datasets, so you don't have to collect the data (or label it), just download!
+
+But if you can't find the dataset that you need? 
+
+Let's see at some tools that can help you in this task:
+
+- [Digitalmethods Tools](https://wiki.digitalmethods.net/Dmi/ToolDatabase) is a set of free web scraping tools developed by Europe's leading Internet Studies groups
+
+- [BeautifulSoup](https://www.dataquest.io/blog/web-scraping-tutorial-python/) is a Python library to scrape webpages
+
+- [Selenium](https://www.pluralsight.com/guides/web-scraping-with-selenium) is a Python library meant to automate browser testing, but it can be perfectly used in combination with BeautifulSoup to set up a powerful web scraper
+
+See also:
+- [10 Best Web Scraping Tools to Extract Online Data](https://www.hongkiat.com/blog/web-scraping-tools/)
+
+Remember that a lot of free access datasets are available for free, for example in the [UCI Dataset](https://archive.ics.uci.edu/ml/datasets.php) website or at the Data science Challenge platform [Kaggle](https://www.kaggle.com/).
+
+## Data Preparation
+
+Data preparation is often the [most time-consuming part of the Data Science project](https://www.infoworld.com/article/3228245/the-80-20-data-science-dilemma.html), so it's crucial to make sure that the right tools are used from you or your team. 
+
+Often the right tools in this phase are the programming languages and their libraries.
+
+The most popular and wide-used libraries are:
+
+- [Scikit-Learn](https://scikit-learn.org/) offers a lot of pre-made data transformations. Get started [here](https://scikit-learn.org/stable/tutorial/basic/tutorial.html).
```

Comment:
There's more see the various Tools section under https://github.com/neomatrix369/awesome-ai-ml-dl/blob/master/things-to-know.md.

Also I didn;t see you mention Fast.ai and other such libraries, these are very good high-level tools

Reply:
Ok but it uses Pytorch and this is addressed later in the guide